pkg/controller/nodeconfig/controller.go

```diff
@@ -0,0 +1,155 @@
+)
+
+const (
+	HandlerName             = "harvester-node-config-controller"
+	ScpoeGlobal             = "Global"
+	ConfigApplied           = "Applied"
+	ConfigAppliedAnnotation = "AppliedConfig"
+)
+
+var toMonitorServices = []string{"NTP", "configFile"}
+
+// ensure every monitor shutdown safely
+var WaitGroup sync.WaitGroup
+
+type Controller struct {
+	ctx      context.Context
+	NodeName string
+
+	NodeConfigs      ctlv1.NodeConfigController
+	NodeConfigsCache ctlv1.NodeConfigCache
+
+	WaitGroup *sync.WaitGroup
+}
+
+func Register(ctx context.Context, nodeName string, nodecfg ctlv1.NodeConfigController, nodes ctlnode.NodeController) (*Controller, error) {
+	ctl := &Controller{
+		ctx:              ctx,
+		NodeName:         nodeName,
+		NodeConfigs:      nodecfg,
+		NodeConfigsCache: nodecfg.Cache(),
+		WaitGroup:        &WaitGroup,
+	}
+
+	monitorNnumbers := len(toMonitorServices)
+
+	monitorModules := make([]interface{}, 0, monitorNnumbers)
+	for _, serviceName := range toMonitorServices {
+		monitorModule := monitor.InitServiceMonitor(ctx, nodecfg, nodes, nodeName, serviceName)
+		monitorModules = append(monitorModules, monitorModule)
+	}
+	monitor.StartsAllMonitors(monitorModules)
+
+	ctl.NodeConfigs.OnChange(ctx, HandlerName, ctl.OnNodeConfigChange)
+	ctl.NodeConfigs.OnRemove(ctx, HandlerName, ctl.OnNodeConfigRemove)
+
+	return ctl, nil
+}
+
+func (c *Controller) OnNodeConfigChange(key string, nodecfg *nodeconfigv1.NodeConfig) (*nodeconfigv1.NodeConfig, error) {
+	confName := strings.Split(key, "/")[1]
+	if nodecfg == nil || confName != c.NodeName || nodecfg.DeletionTimestamp != nil {
+		logrus.Infof("Skip this round (OnChange) with NodeConfigs (%s): %+v", confName, nodecfg)
+		return nil, nil
+	}
+
+	// NTP related handling
+	appliedConfig := nodecfg.ObjectMeta.Annotations[ConfigAppliedAnnotation]
+	ntpConfigHandler := config.NewNTPConfigHandler(nodecfg.Spec.NTPConfig, appliedConfig)
+	updated, err := ntpConfigHandler.DoNTPUpdate(len(nodecfg.Status.NTPConditions) == 0)
+	if err != nil {
+		logrus.Errorf("Update NTP config fail. err: %v", err)
+		return nil, err
+	}
+	if updated {
+		if err := ntpConfigHandler.RestartService(); err != nil {
+			logrus.Errorf("Restart systemd-timesyncd fail. err: %v", err)
+			return nil, err
+		}
+		if err := ntpConfigHandler.UpdateNTPConfigPersistence(); err != nil {
+			logrus.Errorf("Update NTP config to OEM fail. err: %v", err)
+			return nil, err
+		}
+
+		annoValue := generateAnnotationValue(nodecfg.Spec.NTPConfig.NTPServers)
+		bytes, err := json.Marshal(annoValue)
+		if err != nil {
+			logrus.Errorf("Marshal annotation value fail, err: %v", err)
+			return nil, err
+		}
+
+		nodecfgCpy := nodecfg.DeepCopy()
+		if nodecfgCpy.ObjectMeta.Annotations == nil {
+			nodecfgCpy.ObjectMeta.Annotations = make(map[string]string)
+		}
+		nodecfgCpy.ObjectMeta.Annotations[ConfigAppliedAnnotation] = string(bytes)
+		config.UpdateNTPConfigApplied(&nodecfgCpy.Status)
```

Comment:
status is now a subresource. the status updates need to be handled by `c.NodeConfigs.UpdateStatus(nodecfgCpy)`